encoding/marshal.go

```diff
@@ -55,12 +56,74 @@ type (
 	SiaUnmarshaler interface {
 		UnmarshalSia(io.Reader) error
 	}
+)
+
+// An Encoder writes objects to an output stream. It also provides helper
+// methods for writing custom SiaMarshaler implementations. All of its methods
+// become no-ops after the Encoder encounters a Write error.
+type Encoder struct {
+	w   io.Writer
+	buf [8]byte
+	err error
+}
 
-	// An Encoder writes objects to an output stream.
-	Encoder struct {
-		w io.Writer
+// Write implements the io.Writer interface.
+func (e *Encoder) Write(p []byte) (int, error) {
+	if e.err != nil {
+		return 0, e.err
 	}
-)
+	var n int
+	n, e.err = e.w.Write(p)
+	if n != len(p) && e.err == nil {
+		e.err = io.ErrShortWrite
+	}
+	return n, e.err
+}
+
+// WriteByte implements the io.ByteWriter interface.
+func (e *Encoder) WriteByte(b byte) error {
+	if e.err != nil {
+		return e.err
+	}
+	e.buf[0] = b
+	e.Write(e.buf[:1])
+	return e.err
+}
+
+// WriteBool writes b to the underlying io.Writer.
+func (e *Encoder) WriteBool(b bool) error {
+	if b {
+		return e.WriteByte(1)
+	}
+	return e.WriteByte(0)
+}
+
+// WriteUint64 writes a uint64 value to the underlying io.Writer.
+func (e *Encoder) WriteUint64(u uint64) error {
+	if e.err != nil {
+		return e.err
+	}
+	binary.LittleEndian.PutUint64(e.buf[:8], u)
+	e.Write(e.buf[:8])
+	return e.err
+}
+
+// WriteInt writes an int value to the underlying io.Writer.
+func (e *Encoder) WriteInt(i int) error {
+	return e.WriteUint64(uint64(i))
+}
+
+// WritePrefix writes p to the underlying io.Writer, prefixed by its length.
+func (e *Encoder) WritePrefix(p []byte) error {
```

Comment:
Can we rename this to `WriteWithPrefix` ? Every time I come across this method I'm confused since I don't expect it to write the data but only the prefix.

Reply:
agree